File: src/osint_core/schemas/lead.py
"""Pydantic schemas for lead resources."""

import uuid
from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from osint_core.schemas.common import CollectionResponse, SeverityEnum


class LeadTypeEnum(StrEnum):
    """Type of constitutional lead."""

    incident = "incident"
    policy = "policy"


class LeadStatusEnum(StrEnum):
    """Lifecycle status of a lead."""

    new = "new"
    reviewing = "reviewing"
    qualified = "qualified"
    contacted = "contacted"
    retained = "retained"
    declined = "declined"
    stale = "stale"


class LeadResponse(BaseModel):
    """Serialized lead for API responses."""

    model_config = {"from_attributes": True}

    id: uuid.UUID
    lead_type: LeadTypeEnum
    status: LeadStatusEnum

    title: str
    summary: str | None = None
    constitutional_basis: list[str] = Field(default_factory=list)
    jurisdiction: str | None = None
    institution: str | None = None

    severity: SeverityEnum | None = None
    confidence: float | None = None
    dedupe_fingerprint: str

    plan_id: str | None = None
    event_ids: list[uuid.UUID] = Field(default_factory=list)
    entity_ids: list[uuid.UUID] = Field(default_factory=list)
    citations: dict[str, Any] | None = None

    report_id: uuid.UUID | None = None

    first_surfaced_at: datetime
    last_updated_at: datetime
    reported_at: datetime | None = None

    created_at: datetime


class LeadListResponse(CollectionResponse):
    """Paginated list of leads."""

    items: list[LeadResponse]


class LeadUpdateRequest(BaseModel):
    """Request body for updating lead lifecycle state."""

    status: LeadStatusEnum = Field(description="New lifecycle status for the lead")

File: apps/web/src/types/api/lead.ts
import type { CollectionResponse, SeverityEnum } from './common';

export type LeadTypeEnum = 'incident' | 'policy';

export type LeadStatusEnum =
  | 'new'
  | 'reviewing'
  | 'qualified'
  | 'contacted'
  | 'retained'
  | 'declined'
  | 'stale';

export interface LeadResponse {
  id: string;
  lead_type: LeadTypeEnum;
  status: LeadStatusEnum;
  title: string;
  summary: string | null;
  constitutional_basis: string[];
  jurisdiction: string | null;
  institution: string | null;
  severity: SeverityEnum | null;
  confidence: number | null;
  dedupe_fingerprint: string;
  plan_id: string | null;
  event_ids: string[];
  entity_ids: string[];
  report_id: string | null;
  first_surfaced_at: string;
  last_updated_at: string;
  reported_at: string | null;
  created_at: string;
}

export type LeadList = CollectionResponse<LeadResponse>;

export interface LeadUpdateRequest {
  status: LeadStatusEnum;
}
